### PR TITLE
Add `EquiInterleave`

### PR DESCRIPTION
**Pull request by Orace**
This PR addresses #695

### CODE REVIEW

File: MoreLinq/EquiInterleave.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+                if (enumerators.Count == 0)
+                    yield break;
+
+                for (;;)
+                {
+                    var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous();
+
+                    if (isHomogeneous == false)
+                        throw new InvalidOperationException("Input sequences are of different length.");
```
Comment (Orace) [inline MoreLinq/EquiInterleave.cs]:
```suggestion
                        throw new InvalidOperationException("Input sequences are of different lengths.");
```

**Comment by Crossbow78** [comment]
Out of curiosity, what does this add compared to the existing `Interleave` operation, other than an exception with imbalanced sequences?

Wouldn't it be more elegant to make the second `Interleave` overload that accepts a `ImbalancedInterleaveStrategy` argument public, and add a "Throw" strategy to that enumeration? Or at least use it internally, since the enumeration logic should be identical?

**Comment by Orace** [comment]
> Out of curiosity, what does this add compared to the existing `Interleave` operation, other than an exception with imbalanced sequences?
> 
> Wouldn't it be more elegant to make the second `Interleave` overload that accepts a `ImbalancedInterleaveStrategy` argument public, and add a "Throw" strategy to that enumeration? Or at least use it internally, since the enumeration logic should be identical?

The "strategy as parameter" pattern has been discussed for the zip methods. (I will try to find the reference later).
Turns out that "strategy as overload" pattern has been chosen for the zip methods : ZipShortest, ZipLongest and EquiZip.
I just kept this pattern here.

TL; DR; not my choice in first place.

**Comment by atifaziz** [comment]
Closing as #695 was [closed](https://github.com/morelinq/MoreLINQ/issues/695#issuecomment-1399078449) as unplanned for now.
